### PR TITLE
[CodeCompletion][Sema] Allow missing args when solving if the completion location indicates the user may intend to write them later.

### PR DESCRIPTION
**Pull request by nathawes**
`Int` and `String` members should both be prioritized for the completion below despite the missing argument for the first overload since the second argument may just have not been written yet:
```
func foo(a: Int, b: Int) {}
func foo(a: String) {}

foo(a: someBase.<complete here>
```

Similarly, members of type `String` should also be prioritized for the completion below (rather than just `Int`) since the user may intend to use `$1` later in the expression.
```
func bar(a: (Int) -> ()) {}
func bar(a: (String, Int) -> ()) {}

bar { $0.<complete here> }
```

### CODE REVIEW

File: lib/Sema/CSSimplify.cpp [hunk 1]
```diff
@@ -1135,6 +1175,28 @@ class ArgumentFailureTracker : public MatchCallArgumentListener {
   getExtraneousArguments() const {
     return ExtraArguments;
   }
+
+private:
+  bool areMissingArgsInvalid(ConstraintLocator *locator) {
```
Comment (xedin) [inline lib/Sema/CSSimplify.cpp:1180]: Maybe we should convert this into a filter over `MissingArguments` in argument tracker instead and only record the ones which come before the token?

Comment (xedin) [reply lib/Sema/CSSimplify.cpp:1180]: Or even better - ignore missing argument recorded if it comes after the token that way destructor of the `ArgumentFailureTracker` could stay untouched?

File: lib/Sema/CSSimplify.cpp [hunk 2]
```diff
@@ -1771,17 +1832,33 @@ static bool fixMissingArguments(ConstraintSystem &cs, ASTNode anchor,
...
-    return true;
+  // Treat missing anonymous arguments as valid in closures containing the
+  // code completion location, since they may have just not been written yet.
+  if (cs.isForCodeCompletion()) {
+    if (auto *closure = getAsExpr<ClosureExpr>(anchor)) {
+      SourceManager &SM = closure->getASTContext().SourceMgr;
+      SourceRange range = closure->getSourceRange();
+      if (range.isValid() && SM.rangeContainsCodeCompletionLoc(range)) {
+        shouldFix = !closure->hasAnonymousClosureVars() &&
+          (!args.empty() || closure->getInLoc().isValid());
```
Comment (xedin) [inline lib/Sema/CSSimplify.cpp:1851]: Maybe an early return here in reverse situation instead?

**Review by xedin — APPROVED** [review]
Overall LGTM, I have left a couple of comments inline.

File: lib/Sema/CSSimplify.cpp (hunk 2 above)
Comment (nathawes) [reply lib/Sema/CSSimplify.cpp:1851]: Would missing the single argument tuple handling below (need to expand a few lines) be a problem?

Comment (xedin) [reply lib/Sema/CSSimplify.cpp:1851]: Yeah, I think that should be okay.

File: include/swift/Sema/CSFix.h [hunk 3]
```diff
@@ -1178,18 +1178,21 @@ class AllowClosureParamDestructuring final : public ConstraintFix {
                                                 ConstraintLocator *locator);
 };
 
+struct SynthesizedArg {
+  unsigned paramIdx;
+  AnyFunctionType::Param param;
+};
+
```
Comment (nathawes) [inline include/swift/Sema/CSFix.h:1185]: This is no longer necessary (in my original changes it also tracked the index to insert the missing arg at) but I left it in for the slight readability improvement over a `std::pair` (e.g. `paramIdx` vs `first`)

Comment (xedin) [reply include/swift/Sema/CSFix.h:1185]: That works for me, thank you!

**Review by rintaro — APPROVED** [review]

**Comment by nathawes** [comment]
@swift-ci please test

**Comment by nathawes** [comment]
@swift-ci please test and merge

**Pull request merged**